crypto/core_fetch.c

```diff
@@ -124,8 +124,8 @@ void *ossl_method_construct(OSSL_LIB_CTX *libctx, int operation_id,
 {
     void *method = NULL;
 
-    if ((method = mcm->get(NULL, (const OSSL_PROVIDER **)provider_rw,
-                           mcm_data)) == NULL) {
+    if (1 /*(method = mcm->get(NULL, (const OSSL_PROVIDER **)provider_rw,
+                           mcm_data)) == NULL*/) {
```

Comment:
Why is this needed?

Reply:
Because the existing methods in the method store will match although we added a new provider and there would be a better match within that provider.

Only queries that use the preferential property match `?property=x` are affected.

Reply:
So this is a hack to get the tests to pass - but the correct solution is to fix the bug?

Reply:
Yes, exactly. It's also in a commit marked to be dropped. Unfortunately as I said, I'll need some ideas on how to fix the bug properly.

Reply:
Ok, so the problem seems to be that the cache isn't cleared when a new provider is loaded.  That's supposed to happen here:

https://github.com/openssl/openssl/blob/3b836385679504579ee1052ed4b4ef1d9f49fa13/crypto/provider_core.c#L1231

So the question is, why doesn't it happen.  Is `provider_flush_store_cache()` called at all?  If it is, did it fail to actually empty the query caches?

Reply:
I did once wonder why we needed a cache and a store...

Reply:
> I did once wonder why we needed a cache and a store...

It's a *method* store but a *query* cache, i.e the latter saves away the result of specific queries with specific properties.

Of course, if the query cache is deemed unnecessary after all, we could throw it away.

Reply:
> IMO no, the problem is not the cache. That is flushed properly. But the method store is the problem because in case of the optional property match the fetch from the method store will happily succeed although there would be a better match from the activated provider. So yeah, one option would be to not just flush the store cache but also the method store on provider activation. But then, why do we have a separate cache and method store after all?

Do you assume that when someone activates another provider, none of the existing methods are used?

Reply:
I see why you're getting in trouble...  A "maybe" (`?`) property doesn't lend itself well for new providers from which nothing has been fetched before, when there are methods that match the property query with the "maybe" parts ignored.

I don't at all think it's a good idea to clear the method store when a new provider is added, because that will lead to a re-fetch of everything that was in the method store, and we just cleared a huge performance problem related to exactly that last week.

But!  It may be that this change we're discussing here is the right way to go.  After all, the construction of methods will only happen if pre-conditions allow, i.e. if methods from this particular provider aren't already in place according to the provider's opbits.  The call to `ossl_algorithm_do_all()` should therefore be relatively quick the second time around on the same input...

This does forego the *query* cache, though, but considering that its value doesn't seem to be that great, it's possible that the change on this line strikes the right balance, so to say.

I'd suggest running some performance tests on a build of this PR, just to see what difference this particular changes makes.

Reply:
> This does forego the _query_ cache, though, but considering that its value doesn't seem to be that great, it's possible that the change on this line strikes the right balance, so to say.

It shouldn't bypass the cache - that is queried before this code is executed ossl_method_store_cache_get() is called directly from inner_evp_generic_fetch().

Reply:
> I'd suggest running some performance tests on a build of this PR, just to see what difference this particular changes makes.

@paulidale Could you please try to run your msquic performance test environment with this PR in place? It should  hopefully not slowdown things too much.

Reply:
> > This does forego the _query_ cache, though, but considering that its value doesn't seem to be that great, it's possible that the change on this line strikes the right balance, so to say.
> 
> It shouldn't bypass the cache - that is queried before this code is executed: ossl_method_store_cache_get() is called directly from inner_evp_generic_fetch().

Oh right...  I seem to forget that all the time ('cause yeah, when I dove into the performance problem, it took me some time to remind myself how that was done in particular)

Reply:
I don't have a problem emptying the store when a provider is loaded.  Loading/unloading providers is not expected to be a common operation and if it's _heavy_, it will encourage not doing it a lot.

Reply:
> I don't have a problem emptying the store when a provider is loaded. Loading/unloading providers is not expected to be a common operation and if it's _heavy_, it will encourage not doing it a lot.

Refetching all methods is also a heavy operation.  Please let's try to avoid that

Reply:
There is no apparent performance loss in HPS.

Reply:
> There is no apparent performance loss in HPS.

If that's the case, I'll cleanup the last commit so it can be actually merged.